feat: Add PDB Management Operator to ODA Canvas

NoNickeD opened the pull request:
feat: Add PDB Management Operator to ODA Canvas

- Add Pod Disruption Budget management operator
- Supports annotation-based and policy-based PDB creation
- ODA component function awareness (core/security/management)
- Intelligent availability classification for ODA components
- Enterprise observability with metrics, tracing, and audit logging
- Admission webhook validation for policy enforcement
- Maintenance window support for scheduled deployments

Integration Changes:
- Added pdb-management-operator to dockerbuild-config.yaml
- Added Helm values configuration in canvas-oda/values.yaml
- Generated GitHub Actions workflows for automated builds
- Added comprehensive documentation and examples

Operator Features:
- Three enforcement modes: strict, flexible, advisory
- Component function intelligence (core, security, management)
- Policy priority resolution and conflict handling
- Performance optimized for 200+ deployments
- RBAC and security best practices
- Full OpenTelemetry integration

Closes: Integration with TM Forum ODA Canvas ecosystem

LesterThomas commented:
This is the first contribution of a Pod Disruption Budget operator - this will eventually form a closed-loop with the observability (any degradation in performance against SLA will cause the PDB operator to reduced the allowed disruption to pod deployments.)

ferenc-hechler commented:
Hi @NoNickeD,
thanks for your contribution, it is quite comprehensive.   :-)
I am two weeks out of office. 
When I am back, I will take a closer look at this.

ferenc-hechler commented on charts/canvas-oda/values.yaml:
```diff
@@ -364,7 +364,7 @@ pdb-management-operator:
   image: tmforumodacanvas/pdb-management-operator
   version: 1.0.0
-  prereleaseSuffix: -alpha
+  prereleaseSuffix:
```
Please do not remove the prereleaseSuffix until the BDD tests where executed successfully.
For details see https://github.com/tmforum-oda/oda-canvas/blob/main/docs/developer/work-with-dockerimages.md

ferenc-hechler commented:
installation of oda-canvas helm chart fails:
https://github.com/tmforum-oda/oda-canvas/actions/runs/17134852259/job/48625020771?pr=510

![Screenshot_20250821-235902_Firefox~2](https://github.com/user-attachments/assets/a1c30a20-0222-496b-91d1-85ca78dfb966)

ferenc-hechler commented:
Tests are.still failing:
https://github.com/tmforum-oda/oda-canvas/actions/runs/17145802255/job/48642738264?pr=510#step:10:59

ferenc-hechler commented:
It is a security feature, that actions triggered from forks have no access to the Secrets stored in the target GitHub project.
There are some workarounds, but all are not recommended for what we want to do.

So, my suggestion would be to create another PR against a feature-branch, e.g. "feature/510-pdm-management-operator" and merge the changes. This will trigger the necessary docker builds. Then we are good to go with this PR (as long, as there are no changes to the docker image sources.

NoNickeD commented:
Hi @ferenc-hechler,

Done! I've made both changes:

1. Changed prereleaseSuffix from "-alpha" to "alpha" (the hyphen is now removed)
2. Created PR #528 against the `feature/510-build-prerelease-dockerfiles-from-fork` branch as requested

The prerelease Docker images should now build correctly once you merge PR #528. The image tag will be `1.0.0-alpha` (with single hyphen).

Thanks for your patience and guidance on this!

ferenc-hechler commented on source/operators/pdb-management/examples/demo/scripts/demo-auth-service.sh:
```diff
@@ -0,0 +1,37 @@
+
+  # Start log monitoring in background
+  echo "2⃣ Starting log monitor..."
+  kubectl logs -n canvas deployment/pdb-management-controller-manager -f --tail=0 | \
```
Using the helm chart oda-canvas the operator name is `canvas-pdb-management-operator` instead of `pdb-management-controller-manager`.

Also in other places the deployment name `pdb-management-controller-manager`is used and should be replaced with `canvas-pdb-management-operator`

ferenc-hechler commented:
After manually fixing the deployment name of the pdb-mgmt operator and manually creating the pdb-demo namespace the script demo-auth-service.sh was successfully executed:

```
🚀 Starting Auth Service Demo...
----------------------------------------
1⃣ Cleaning up any existing deployment...
2⃣ Starting log monitor...
3⃣ Applying auth-service deployment...
Found 2 pods, using pod/canvas-pdb-management-operator-68cdfd86b7-n79l4
deployment.apps/auth-service created
----------------------------------------
4⃣ Final Results:
NAME               MIN-AVAILABLE   DISRUPTIONS-ALLOWED
auth-service-pdb   90%             0
```

Also I can see the generated PDB:

```
$ kubectl get pdb -n pdb-demo
NAME               MIN AVAILABLE   MAX UNAVAILABLE   ALLOWED DISRUPTIONS   AGE
auth-service-pdb   90%             N/A               1                     6m27s
```

For me the smoke tests are sufficient.

As there are currently no PDB-Management-Operator Tests in the BDD a unsuccessful deployment of the operator (e.g. Image-Pull-Error) is not detected by the BDD tests.

After fixing the doc to match the current operator name `canvas-pdb-management-operator` I can approve this PR

ferenc-hechler reviewed (approved):
Approved.

Please remove the prereleaseSuffix, them we can merge this PR

NoNickeD commented:
Hi @ferenc-hechler,

Done! I've removed the prereleaseSuffix as requested.

The commit was made with `[skip tests]` prefix to avoid re-running the BDD tests.

All PR checks should now pass and the PR is ready for merging. Once merged to main, the release Docker image `tmforumodacanvas/pdb-management-operator:1.0.0` will be automatically built.

Thank you for your thorough review and testing!

ferenc-hechler commented:
@NoNickeD : thanks for your great contribution!   
<img width="78" height="78" alt="image" src="https://github.com/user-attachments/assets/45cbd383-f359-44d6-8e77-6ad6a0d5e76a" />

It is the first GO operator and will be a template for others to look how to implement advanced operators.

To ensure the continuing functionality of the operator it would be great, if you could create a BDD tests, which checks the base functionality of the PDB Mgmt Operator
